File: src/ModelBuilder/Variables.jl
module Variables

export 
    define_model_variables_1ph_NL_t_in_Tset,
    define_model_variables_1ph_NL_t_in_Tset

using Parameters: @unpack, @pack!
using JuMP

#region define_model_variables_1ph_NL_t_in_Tset
"""
    define_model_variables_1ph_NL_t_in_Tset(modelDict; Tset=nothing)

Define the optimization model variables for a single-phase network with nonlinear loads over a given time set.

This function defines the decision variables for the optimization model stored in `modelDict`. 
It handles different sets of variables, including power flows, voltages, and state of charge (SOC) for batteries.
"""
function define_model_variables_1ph_NL_t_in_Tset(modelDict; Tset=nothing)
    @unpack model, data = modelDict;
    if Tset === nothing
        Tset = data[:Tset]
    end

    # @show Tset

    @unpack Nset, Lset, Dset, Bset, PSubsMax_kW, kVA_B = data
    PSubsMax_pu = PSubsMax_kW / kVA_B

    # Define all variables as before, using the data parsed
    @variable(model, PSubsMax_pu >= P_Subs[t in Tset] >= 0)
    # Define variables over the set of branches Lset and time periods Tset
    @variable(model, P[(i, j) in Lset, t in Tset] <= PSubsMax_pu, base_name = "P")
    @variable(model, Q[(i, j) in Lset, t in Tset], base_name = "Q")
    @variable(model, l[(i, j) in Lset, t in Tset] >= 0, base_name = "l")

    @variable(model, v[j in Nset, t in Tset], base_name = "v")
    @variable(model, q_D[j in Dset, t in Tset], base_name = "q_D")
    @variable(model, q_B[j in Bset, t in Tset], base_name = "q_B")
    @variable(model, P_c[j in Bset, t in Tset], base_name = "P_c")
    @variable(model, P_d[j in Bset, t in Tset], base_name = "P_d")

    # Since SOC constraints for the t-th hour require SOCs of the previous time-steps i.e. B[j, t-1], we need to ensure that they too (in addition to B[j, t]) are defined. This is important in the case where Tset = [t0] (i.e. only one time-step) and t0 != 1 (for t0=1, B[j, t-1] is not required as it is separately taken from B0 stored in data)
    if !(1 ∈ Tset)
        t0m1 = Tset[1] - 1 # assuming Tset is sorted, which it should be
        Tset_appended = [t0m1; Tset]
        @variable(model, B[j ∈ Bset, t ∈ Tset_appended], base_name = "B")
    elseif 1 ∈ Tset
        @variable(model, B[j ∈ Bset, t ∈ Tset], base_name = "B")
    else
        error("Tset must contain at least one element")
    end

    # @variable(model, B[j in Bset, t in Tset], base_name = "B")

    @pack! modelDict = model
    return modelDict
end
#endregion

#region define_model_variables_1ph_NL_t_in_Tset
"""
    define_model_variables_1ph_NL_t_in_Tset(modelDict; Tset=nothing)

Define the optimization model variables for a single-phase network with nonlinear loads over a given time set.

This function defines the decision variables for the optimization model stored in `modelDict`. 
It handles different sets of variables, including power flows, voltages, and state of charge (SOC) for batteries.
"""
function define_model_variables_1ph_NL_t_in_Tset(modelDict; Tset=nothing)
    @unpack model, data = modelDict
    if Tset === nothing
        Tset = data[:Tset]
    end

    # @show Tset

    @unpack Nset, Lset, Dset, Bset, PSubsMax_kW, kVA_B = data
    PSubsMax_pu = PSubsMax_kW / kVA_B

    # Define all variables as before, using the data parsed
    @variable(model, PSubsMax_pu >= P_Subs[t in Tset] >= 0)
    # Define variables over the set of branches Lset and time periods Tset
    @variable(model, P[(i, j) in Lset, t in Tset] <= PSubsMax_pu, base_name = "P")
    @variable(model, Q[(i, j) in Lset, t in Tset], base_name = "Q")

    @variable(model, v[j in Nset, t in Tset], base_name = "v")
    @variable(model, q_D[j in Dset, t in Tset], base_name = "q_D")
    @variable(model, q_B[j in Bset, t in Tset], base_name = "q_B")
    @variable(model, P_c[j in Bset, t in Tset], base_name = "P_c")
    @variable(model, P_d[j in Bset, t in Tset], base_name = "P_d")

    # Since SOC constraints for the t-th hour require SOCs of the previous time-steps i.e. B[j, t-1], we need to ensure that they too (in addition to B[j, t]) are defined. This is important in the case where Tset = [t0] (i.e. only one time-step) and t0 != 1 (for t0=1, B[j, t-1] is not required as it is separately taken from B0 stored in data)
    if !(1 ∈ Tset)
        t0m1 = Tset[1] - 1 # assuming Tset is sorted, which it should be
        Tset_appended = [t0m1; Tset]
        @variable(model, B[j ∈ Bset, t ∈ Tset_appended], base_name = "B")
    elseif 1 ∈ Tset
        @variable(model, B[j ∈ Bset, t ∈ Tset], base_name = "B")
    else
        error("Tset must contain at least one element")
    end

    # @variable(model, B[j in Bset, t in Tset], base_name = "B")

    @pack! modelDict = model
    return modelDict
end
#endregion

end

File: src/ModelBuilder_py/Variables.py
from pyomo.environ import Var, NonNegativeReals

def define_model_variables_1ph_NL_t_in_Tset(model, data, Tset=None):
    if Tset is None:
        Tset = data['Tset']

    # Unpack the data
    Nset = data['Nset']
    Lset = data['Lset']
    Dset = data['Dset']
    Bset = data['Bset']
    PSubsMax_kW = data['PSubsMax_kW']
    kVA_B = data['kVA_B']
    PSubsMax_pu = PSubsMax_kW / kVA_B

    # Define all variables as before, using the data parsed
    model.P_Subs = Var(Tset, bounds=(0, PSubsMax_pu))

    # Define variables over the set of branches Lset and time periods Tset
    model.P = Var([(i, j, t) for (i, j) in Lset for t in Tset], bounds=(None, PSubsMax_pu), domain=NonNegativeReals)
    model.Q = Var([(i, j, t) for (i, j) in Lset for t in Tset], domain=NonNegativeReals)
    model.l = Var([(i, j, t) for (i, j) in Lset for t in Tset], domain=NonNegativeReals)

    model.v = Var([(j, t) for j in Nset for t in Tset], domain=NonNegativeReals)
    model.q_D = Var([(j, t) for j in Dset for t in Tset], domain=NonNegativeReals)
    model.q_B = Var([(j, t) for j in Bset for t in Tset], domain=NonNegativeReals)
    model.P_c = Var([(j, t) for j in Bset for t in Tset], domain=NonNegativeReals)
    model.P_d = Var([(j, t) for j in Bset for t in Tset], domain=NonNegativeReals)

    model.B = Var([(j, t) for j in Bset for t in Tset], domain=NonNegativeReals)

    # Handle SOC constraints for previous time steps
    if 1 not in Tset:
        t0m1 = min(Tset) - 1  # Assuming Tset is sorted
        model.B_t0m1 = Var([(j, t0m1) for j in Bset], domain=NonNegativeReals)

    return model
